sliceanddice/host/src/main/java/com/novoda/sliceanddice/MainActivity.kt

```diff
@@ -0,0 +1,131 @@
+package com.novoda.sliceanddice
+
+import android.content.Intent
+import android.net.Uri
+import android.os.Bundle
+import android.support.constraint.ConstraintSet
+import android.support.v7.app.AppCompatActivity
+import android.transition.TransitionManager
+import android.view.View
+import android.widget.ArrayAdapter
+import android.widget.Toast
+import androidx.slice.SliceManager
+import com.novoda.sliceshost.R
+import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.android.synthetic.main.settings_card.*
+
+class MainActivity : AppCompatActivity() {
+
+    private lateinit var sliceManager: SliceManager
+    private var pendingSliceUri: Uri? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        sliceManager = SliceManager.getInstance(this)
+        sliceSelector.adapter = ArrayAdapter<SliceChoice>(this, android.R.layout.simple_list_item_1, sliceChoices)
+
+        transitionUiTo(UiState.Empty, animate = false)
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        if (providerAppNotInstalled(packageManager, baseSliceUri.authority)) {
+            showMissingProviderDialog(this, { finish() }, baseSliceUri)
+            return
+        }
+
+        sliceSelector.onItemSelectedListener = sliceSelectionListener({ _, _, selectedPosition, _ ->
+            onSliceSelectionChanged(sliceChoices[selectedPosition].uri)
+        })
+
+        sliceSelector.setSelection(0)
+    }
+
+    private fun onSliceSelectionChanged(selectedSliceUri: Uri) {
+        tryShowingSlice(selectedSliceUri)
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        handleSlicePermissionActivityResult(requestCode, onSlicePermissionResult = {
+            if (pendingSliceUri == null) {
+                throw IllegalArgumentException("The slice URI to bind to cannot be null")
+            }
+            tryShowingSlice(pendingSliceUri!!)
+            pendingSliceUri = null
+        })
+    }
+
+    private fun tryShowingSlice(sliceUri: Uri) {
+        if (sliceManager.missingPermission(sliceUri, appName = getString(R.string.app_name))) {
+            transitionUiTo(UiState.NeedPermission)
+            permissionCta.setOnClickListener {
+                pendingSliceUri = sliceUri
+                sliceManager.requestPermission(sliceUri, this)
+            }
+        } else {
+            startObservingSliceLiveData(sliceUri)
+        }
+    }
+
+    private fun startObservingSliceLiveData(sliceUri: Uri) {
+        transitionUiTo(UiState.SliceContent)
+
+        val slice = sliceManager.bindSlice(sliceUri)
```

Comment:
Nah we're fine, it's `SliceView`'s problem in case. We do extensive checks on the URI anyway, the only case where the slice could be `null` in this sample is if the provider emitted `null`, which it never does. I would not do defensive programming against things we are sure are not going to happen. If someone else changes the sample, it's their responsibility to make things work for them

Reply:
Makes sense, lets keep the sample as clean as possible to better _sample_ things